keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -589,6 +590,29 @@ associativeList.all(e, e.key1 == 'a' && e.key2 == 'b' && e.val == 100)
 associativeList.all(e, e.val == 100)
 ```
 
+### Resource constraints
+
+For Beta, per-request execution time will be constrained via a context-passed timeout. For Beta, we are looking at a range between 100 milliseconds and 1 second for the timeout. We want to see how
+CEL performs in benchmarks once our Beta optimizations are in place before deciding what specific 
+duration to set the timeout to. If the timeout is  exceeded, the error message will include how
+long each expression took to evaluate.
+
+We will provide time and space benchmarks to show what can be expected in typical and worst-case
+scenarios for CEL expressions, and that for most use cases, the timeout alone will be sufficient.
+The alternatives listed below would be more aimed at malformed/malicious expressions.
+
+If the context timeout proves to be insufficient, we have other alternatives to explore, such as
+limiting nested list comprehensions to a depth of 2. According to the performance section of
+[the CEL spec](https://github.com/google/cel-spec/blob/master/doc/langdef.md#performance), the
+time complexity of these operations are all O(n), so by limiting nesting to 2 deep, we limit
+expression complexity to O(n<sup>2</sup>). We can limit regular expression complexity
+as well.
+
+CEL also provides a [cost subsystem](https://github.com/google/cel-go/blob/dfef54b359b05532fb9695bc88937aa8530ab055/cel/program.go#L309) that could be used in the future,
+but the cost subsystem would need to know the length of any relevant lists in order to be useful. 
+That information can be supplied using `maxLength`, but this is an optional field, and if not
+passed, CEL would not be able to provide a useful figure.
```

Comment:
I'm not sure if it's useful, but we could set a max length of a list based on the 3MB request limit and the size of the most compact encoding possible:
* int / float lists = 3*1024*1024 / 2 (`0,0,0,0,0,0,0,0,0,...` encoding)
* string lists = 3*1024*1024 / 3 (`"","","","","",...` encoding)
* list lists = 3*1024*1024 / 3 (`[],[],[],[],[],...` encoding)
* object lists = 3*1024*1024 / 3 (`{},{},{},{},{},...` encoding)

I'm not sure if passing such big list sizes to CEL would make the cost subsystem go nuts or not, but we could check

Reply:
Yeah, I'll give passing lists of that size through CEL a shot and see what happens. I think that list might work well to provide `maxLength` with a default value, though I'm concerned that there will be a large discrepancy between what we'd estimate resource usage to be versus what we'd see in practice if we actually ran the expressions. I suspect any solution that works without `maxLength` being explicitly set will run into the same issue, though.